hivemind/client/switch_moe.py

```diff
@@ -0,0 +1,177 @@
+            self.backward_timeout, self.detect_anomalies, self.allow_zero_outputs, self.info,
+            *nested_flatten(((input, *args), kwargs)))
+        # ^-- multiple tensors of shape [batch_size, max_experts, ...output_shape]
+
+        batch_utilization = self._compute_batch_utilization(chosen_experts, expert_mask)
+        self.grid_utilization = \
+            self.utilization_alpha * self.grid_utilization + (1 - self.utilization_alpha) * batch_utilization
+
+        # compute expert probabilities as product across grid dimensions
+        expert_probs = self.compute_expert_scores(grid_softmax, chosen_experts)
+        masked_logits = torch.full((1,), float('-inf'), device=expert_probs.device, dtype=expert_probs.dtype)
+        expert_probs = torch.where(expert_mask, expert_probs, masked_logits)
+
+        # multiply outputs by expert probabilities
+        averaged_outputs_flat = [
+            (expert_probs[..., None] * tensor.flatten(start_dim=2)).view(tensor.shape).sum(dim=1)
+            for tensor in expert_outputs]  # ^-- multiply by softmax weights along first 2 axes
+
+        packed_outputs = nested_pack(averaged_outputs_flat, self.info['outputs_schema'])
+
+        # balancing loss: fraction of examples routed to each expert,
+        # fraction of probability assigned to each expert (both across all experts)
+        # want both to have equal allocation
+
+        balancing_loss = torch.stack([torch.mean(dim_softmax.mean(0) * dim_utilization) * (dim_size ** 2)
+                                      for dim_softmax, dim_utilization, dim_size in
+                                      zip(grid_softmax, self.grid_utilization, self.beam_search.grid_size)]).sum()
+
+        # residual connection
+        if isinstance(packed_outputs, torch.Tensor):
+            packed_outputs = packed_outputs + input
+        else:
+            packed_outputs[0] = packed_outputs[0] + input
+
+        return packed_outputs, balancing_loss
+
+    @torch.no_grad()
+    def _compute_batch_utilization(self, batch_experts, expert_mask):
+        batch_utilization = [torch.zeros((dim_size,), dtype=self.grid_utilization.dtype,
+                                         device=self.grid_utilization.device)
+                             for dim_size in self.beam_search.grid_size]
+
+        # out of chosen_experts, select those for which expert_mask is True
+        for (sample_idx, expert_idx) in expert_mask.nonzero().numpy():
+            expert = batch_experts[sample_idx][expert_idx]
+            expert_indices = expert.uid[len(self.beam_search.uid_prefix):]
+            expert_indices = list(map(int, expert_indices.split(UID_DELIMITER)))
+
+            for dim_index, dim_utilization in zip(expert_indices, batch_utilization):
+                dim_utilization[dim_index] += 1
+
+        return torch.cat([
+            torch.nn.functional.normalize(dim_utilization, p=1, dim=0)
+            for dim_utilization in batch_utilization
+        ])
```

Comment:
Probably suboptimal, any advice on speeding that up would be welcome

Reply:
We can probably use .scatter_add_ to avoid loops. Mind if i create an issue?